src/test/regress/expected/functional_deps_optimizer.out

```diff
@@ -0,0 +1,1782 @@
+               ^
+-- group by right table's (composite) primary key (OK)
+SELECT aic.changed
+FROM articles AS a JOIN articles_in_category AS aic ON a.id = aic.article_id
+WHERE aic.category_id in (14,62,70,53,138)
+GROUP BY aic.category_id, aic.article_id;
+ changed 
+---------
+(0 rows)
+
+-- group by right table's partial primary key (fail)
+SELECT aic.changed
+FROM articles AS a JOIN articles_in_category AS aic ON a.id = aic.article_id
+WHERE aic.category_id in (14,62,70,53,138)
+GROUP BY aic.article_id;
+ERROR:  column "aic.changed" must appear in the GROUP BY clause or be used in an aggregate function
+LINE 1: SELECT aic.changed
+               ^
+-- example from documentation
+CREATE TEMP TABLE products (product_id int, name text, price numeric);
+CREATE TEMP TABLE sales (product_id int, units int);
+-- OK
+SELECT product_id, p.name, (sum(s.units) * p.price) AS sales
+    FROM products p LEFT JOIN sales s USING (product_id)
+    GROUP BY product_id, p.name, p.price;
+ product_id | name | sales 
+------------+------+-------
+(0 rows)
+
+-- fail
+SELECT product_id, p.name, (sum(s.units) * p.price) AS sales
+    FROM products p LEFT JOIN sales s USING (product_id)
+    GROUP BY product_id;
+ERROR:  column "p.name" must appear in the GROUP BY clause or be used in an aggregate function
+LINE 1: SELECT product_id, p.name, (sum(s.units) * p.price) AS sales
+                           ^
+ALTER TABLE products ADD PRIMARY KEY (product_id);
+-- OK now
+SELECT product_id, p.name, (sum(s.units) * p.price) AS sales
+    FROM products p LEFT JOIN sales s USING (product_id)
+    GROUP BY product_id;
+ product_id | name | sales 
+------------+------+-------
+(0 rows)
+
+-- OK, test GPDB case
+set enable_groupagg = off;
+set gp_eager_two_phase_agg = on;
+SELECT count(distinct name), price FROM products GROUP BY product_id;
+ count | price 
+-------+-------
+(0 rows)
+
+create table funcdep1(a int primary key, b int, c int, d int);
+create table funcdep2(a int, b int, c int, d int);
+NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'a' as the Greenplum Database data distribution key for this table.
+HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
+insert into funcdep1 values(1,1,1,1);
+insert into funcdep1 values(2,1,1,1);
+insert into funcdep1 values(3,1,1,1);
+insert into funcdep2 values(1,1,1,1);
+explain (costs off) select sum(t2.a), t1.a, t1.b, t1.c from funcdep1 t1 join funcdep2 t2 on t1.b = t2.b group by t1.a;
+                            QUERY PLAN                            
+------------------------------------------------------------------
+ GroupAggregate
+   Group Key: funcdep1.a, funcdep1.b, funcdep1.c
+   ->  Sort
+         Sort Key: funcdep1.a, funcdep1.b, funcdep1.c
+         ->  Hash Join
+               Hash Cond: (funcdep1.b = funcdep2.b)
+               ->  Gather Motion 3:1  (slice1; segments: 3)
```

Comment:
How good is it that ORCA now generates a non-optimal plan where all the work happens on the coordinator?

Reply:
I guess this topic is out of scope of the task. The task was to make GPORCA support this kind of queries.
Their further optimization is another task, pretty large in scope I think.

Reply:
On 7X plan for this query is optimal

Reply:
Then I guess if we integrate this change to 7X, 
```
explain (costs off) select sum(t2.a), t1.a, t1.b, t1.c from funcdep1 t1 join funcdep2 t2 on t1.b = t2.b group by t1.a
```
 will also produce optimal plan, because it will be transformed to
 ```
explain (costs off) select sum(t2.a), t1.a, t1.b, t1.c from funcdep1 t1 join funcdep2 t2 on t1.b = t2.b group by t1.a, t1.b, t1.c;
```

Reply:
Without inserted data 6X version produces the same plan as 7X version.